### PR TITLE
fixes #29: implement an approximate size function

### LINKED ISSUE #29: replace MallocSizeOf with an approximate size function
Due to in-tree/out-of-tree version issues, it's not really feasible to have this crate implement MallocSizeOf. Instead, it can provide a function that gives the approximate memory usage of each Cascade.

### PR DESCRIPTION
**Pull request by mozkeeler**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -71,10 +66,15 @@ impl BitVector {
         };
         test_value > 0
     }
+
+    /// Determine the approximate amount of memory used by this BitVector.
+    /// See the caveat in Cascade::approximate_size.
+    fn approximate_size(&self) -> usize {
+        self.bytes.len() + size_of::<usize>()
```
Comment (jschanck) [inline src/lib.rs]: nit: either the `usize` is negligible and we don't care about it, or we also care about the Vec struct and padding. I would either go with `self.bytes.len()` or `size_of::<BitVector>() + self.bytes.len()`, and I guess I lean towards the latter.

File: src/lib.rs [hunk 2]
```diff
@@ -209,6 +208,16 @@ impl Bloom {
...
+
+    /// Determine the approximate amount of memory used by this Bloom.
+    /// See the caveat in Cascade::approximate_size.
+    fn approximate_size(&self) -> usize {
+        size_of::<u8>()
+            + size_of::<u32>()
+            + size_of::<u32>()
+            + self.bit_vector.approximate_size()
+            + size_of::<HashAlgorithm>()
+    }
```
Comment (jschanck) [inline src/lib.rs]: Here as well, `size_of::<Bloom>() + self.bit_vector.approximate_size()`

File: src/lib.rs [hunk 3]
```diff
@@ -317,6 +325,22 @@ impl Cascade {
...
+    /// the bloom filters are large enough to dominate the overall allocated
+    /// size.
+    pub fn approximate_size(&self) -> usize {
+        self.filter.approximate_size()
+            + self
+                .child_layer
+                .as_ref()
+                .map_or(0, |child_layer| child_layer.approximate_size())
+            + self.salt.as_ref().map_or(0, |salt| salt.len())
+            + size_of::<bool>()
```
Comment (jschanck) [inline src/lib.rs]: and might as well throw in a `size_of::<Cascade>` for good measure.

**Review by jschanck — APPROVED** [review]
This will do just fine. But I proposed some tweaks below.

File: src/lib.rs (hunk 3 above)
Comment (mozkeeler) [reply src/lib.rs]: Right, so my understanding of rust is `size_of::<Cascade>()` will include `size_of::<Bloom>()` and `size_of::<BitVector>()`. Doing it this way simplifies things in a way, in that we can basically boil this all down to `size_of::<Cascade>()` plus the lengths of the filter bytes and salts.

**Review by jschanck — APPROVED** [review]
Nice!

**Pull request merged**